src/main/java/org/opentripplanner/routing/algorithm/raptor/router/street/AccessEgressFilter.java

```diff
@@ -29,16 +34,61 @@
     switch (streetMode) {
       case CAR_PICKUP:
       case CAR_TO_PARK:
-        return filterByDistanceOfNearestStop(
+        return filterByClosestStationsPerPatterns(
             nearbyStops,
-            request.accessEgressFilterMinimumDistanceMeters,
-            request.accessEgressFilterDistanceFactor
+            request.maxNumberOfStationsForCarAccessEgress,
+            request.rctx.graph.index
         );
       default:
         return nearbyStops;
     }
   }
 
+  /**
+   * Returns the closest stations for each TripPattern.
+   */
+  private static Collection<NearbyStop> filterByClosestStationsPerPatterns(
+      Collection<NearbyStop> nearbyStops, int maxNumberOfStations, GraphIndex graphIndex
+  ) {
+    if (nearbyStops.isEmpty()) { return nearbyStops; }
+
+    // Because order is kept using these stream operators, we will have unique stations by distance
+    List<Station> stationsByDistance = nearbyStops
+        .stream()
+        .sorted()
+        .filter(s -> s.stop instanceof Stop)
+        .map(s -> ((Stop) s.stop).getParentStation())
```

Comment:
In our feed we don't have parent stations so they are null. This leads to a list with a single null value being returned here.

If there is no parent station then simply the stop itself ought to be user, isn't it?

Reply:
I didn't really think of that, as I am so used to stops always having a station.

We probably should have the code handle this, but I think the code here and in some other places too, would be a lot cleaner if we knew all the stops had stations (maybe by adding stations for all stops in the gtfs import process).

Reply:
We should also think a bit about how it should be mixed, because finding the 5 closest stops can be very different from finding the 5 closest stations.